### PR TITLE
Performance metrics for AquaVM  based on the `tracing` crate

### PR DESCRIPTION
**Pull request by monoid**
Current version outputs JSON traces with execution times of important parts of the AquaVM.

### CODE REVIEW

File: air-interpreter/src/logger.rs [hunk 1]
```diff
@@ -15,7 +15,6 @@
  */
 
 use air_log_targets::TARGET_MAP;
-
```
Comment (ValeryAntopol) [inline air-interpreter/src/logger.rs:18]: Its nice to have internal/external crates imports separated, so lets keep the blank line

File: air-interpreter/src/wasm_bindgen.rs [hunk 2]
```diff
@@ -41,8 +41,10 @@ pub const DEFAULT_LOG_LEVEL: LevelFilter = LevelFilter::Trace;
 pub fn main() {
     // it's necessary to initialize it with the minimal allowed log level,
     // because otherwise it's impossible to set less level than used during initialization.
+
```
Comment (ValeryAntopol) [inline air-interpreter/src/wasm_bindgen.rs:44]: Looks like we forgot to remove this compilation target. Maybe we can just remove it in this PR, @mikevoronov what do you think?

**Review by ValeryAntopol — COMMENTED** [review]
utACK
But i didn't check that set of skipped/traced functions is correct,  @mikevoronov will do a better job looking at it.

File: air-interpreter/src/wasm_bindgen.rs (hunk 2 above)
Comment (monoid) [reply air-interpreter/src/wasm_bindgen.rs:44]: I think removing this file is too important to be merged with other PR.

File: air/src/execution_step/air/call/resolved_call.rs [hunk 3]
```diff
@@ -129,7 +130,11 @@ impl<'i> ResolvedCall<'i> {
             tetraplets,
         } = self.resolve_args(exec_ctx)?;
 
-        let serialized_tetraplets = serde_json::to_string(&tetraplets).expect("default serializer shouldn't fail");
+        let serialized_tetraplets = {
+            let span = tracing::span!(tracing::Level::INFO, "serde_json::to_string(tetraplets)");
```
Comment (monoid) [inline air/src/execution_step/air/call/resolved_call.rs:134]: Such construction now exist in few places, and it looks like a great application for a macros.  I haven't found proper place for it.

File: tools/cli/air-trace/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+use std::time::Duration;
+
+pub(crate) fn parse_tracing_duration(input: &str) -> Result<Duration, anyhow::Error> {
+    for (suffix, scale) in [("ns", 1e-9), ("µs", 1e-6), ("ms", 1e-3), ("s", 1e0)] {
+        if let Some(num_str) = input.strip_suffix(suffix) {
+            if let Ok(num) = num_str.parse::<f64>() {
+                return Ok(Duration::from_secs_f64(num * scale));
+            } else {
+                break;
+            }
```
Comment (mikevoronov) [inline tools/cli/air-trace/src/utils.rs:25]: Let's use `match`

File: tools/cli/air-trace/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+use std::time::Duration;
+
+pub(crate) fn parse_tracing_duration(input: &str) -> Result<Duration, anyhow::Error> {
+    for (suffix, scale) in [("ns", 1e-9), ("µs", 1e-6), ("ms", 1e-3), ("s", 1e0)] {
+        if let Some(num_str) = input.strip_suffix(suffix) {
+            if let Ok(num) = num_str.parse::<f64>() {
+                return Ok(Duration::from_secs_f64(num * scale));
+            } else {
+                break;
```
Comment (mikevoronov) [inline tools/cli/air-trace/src/utils.rs:24]: Can we use `return Err` here?

File: tools/cli/air-trace/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    return Err(anyhow::anyhow!("malformed duration {:?}", input));
+}
+
+pub(crate) fn unix_timestamp_now() -> u64 {
+    use std::time::SystemTime;
+
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("time befor Unix epoch")
```
Comment (mikevoronov) [inline tools/cli/air-trace/src/utils.rs:36]: befor**e**

File: tools/cli/air-trace/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use std::time::Duration;
+
+pub(crate) fn parse_tracing_duration(input: &str) -> Result<Duration, anyhow::Error> {
```
Comment (mikevoronov) [inline tools/cli/air-trace/src/utils.rs:18]: Are there any dedicated crate for that? Looks like smth like https://docs.rs/parse_duration/latest/parse_duration/ could suit (but I haven't look much to its API).

File: air/src/execution_step/air/call/resolved_call.rs (hunk 3 above)
Comment (mikevoronov) [reply air/src/execution_step/air/call/resolved_call.rs:136]: Yep, I'd agree that it's a good application for macros.

File: air/src/farewell_step/outcome.rs [hunk 8]
```diff
@@ -100,7 +113,8 @@ fn populate_outcome_from_contexts(
 }
 
 /// Deduplicate values in a supplied vector.
-fn dedup<T: Eq + Hash>(mut vec: Vec<T>) -> Vec<T> {
+#[tracing::instrument]
```
Comment (mikevoronov) [inline air/src/farewell_step/outcome.rs:116]: I don't need we actually need to instrument it, other functions will dominate over this deduplication.

File: air/src/runner.rs [hunk 9]
```diff
@@ -64,7 +64,13 @@ fn execute_air_impl(
...
     // match here is used instead of map_err, because the compiler can't determine that
     // they are exclusive and would treat exec_ctx and trace_handler as moved
-    match air.execute(&mut exec_ctx, &mut trace_handler) {
+    match {
+        // create a span for tracing the whole air.execute and its handling
+        let execute_span = tracing::span!(tracing::Level::INFO, "execute");
+        let _enter = execute_span.enter();
+
+        air.execute(&mut exec_ctx, &mut trace_handler)
+    } {
```
Comment (mikevoronov) [inline air/src/runner.rs:73]: Let's add a new variable and then match on it.

File: tools/cli/air-trace/src/run.rs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    #[clap(name = "--plain")]
+    PlainData(self::data::plain::PlainDataArgs),
+}
+
+pub(crate) fn run(args: Args) -> anyhow::Result<()> {
+    let tracing_json = (!args.json) as u8;
+    init_tracing(args.tracing_params.clone(), tracing_json);
+
+    let current_peer_id = args.current_peer_id.as_deref().unwrap_or("some_id");
+    let mut runner = if cfg!(not(feature = "wasm")) || args.native {
```
Comment (mikevoronov) [inline tools/cli/air-trace/src/run.rs:73]: Suggest to introduce a dedicated function for getting runner

File: tools/cli/air-trace/src/run.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+use air_test_utils::CallResults;
+
+use anyhow::Context as _;
+use clap::{Parser, Subcommand};
+use std::path::{Path, PathBuf};
+
+#[derive(Parser, Debug)]
+#[clap(about = "Run AIR script with AquaVM")]
+pub(crate) struct Args {
```
[the remark this answers is not in the document]
Comment (monoid) [reply tools/cli/air-trace/src/run.rs:31]: Newlines separate logical groups of arguments.  Perhaps, one should refactor out them into flattened structs.

File: air/src/runner.rs (hunk 9 above)
Comment (monoid) [reply air/src/runner.rs:73]: It will be solved by the macro.

File: tools/cli/air-trace/src/utils.rs (hunk 7 above)
Comment (monoid) [reply tools/cli/air-trace/src/utils.rs:18]: I'v tried several ones, but one does not support floats, another does not support suffixes used by the output format, etc.

**Review by mikevoronov — APPROVED** [review]
tested ACK

**Pull request merged**